Add new-style string formatting option and callable option to `fmt` in `Axes.bar_label()`.

stefmolin opened the pull request:
## PR Summary
Addresses #23689

Enhancement to `Axes.bar_label()` to make it possible to format values with f-strings.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [X] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] New features are documented, with examples if plot related.
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

stefmolin commented:
Example usage:

```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
bar_container = ax.barh(['c', 'b', 'a'], [1_000, 5_000, 7_000])
ax.bar_label(bar_container, label_type='center', fmt='{:,.0f}')
```

<img width="362" alt="Screen Shot 2022-08-20 at 6 44 41 PM" src="https://user-images.githubusercontent.com/24376333/185768491-512c4e38-b8a2-4393-a72e-ea7eae890d1d.png">

story645 commented:
I think this is totally reasonable given we allow fstrings directly and indirectly for most of the other labeling, so I'm +1 on this.

tacaswell commented:
In general we have not tried to guess which version of format strings the user passed.  In the test used here `"label {:f}"` and `"{%f}"` would both fail the wrong way.  

This did come up when we added this in 2019 https://github.com/matplotlib/matplotlib/pull/15602#discussion_r347118557 and we opted for "old style" at the time.

I do not think we should go down the route of guessing which one the user passed (in addition to the two examples above `"a: %d b: {:d}"` I think will correctly format under either scheme and we have no way to tell which the user meant (I'm sure there are more realistic examples that have the same problem)).   We have not done it elsewhere in the library (this is why we have a `FormatStrFormatter` (old style) and `StrMethodFormatter` (new style) classes).

I think our options are

1. add a flag to control how we interpret `fmt` 
2. add a second (conflicting) keyword for newsytle format (if both are passed error, if neither is passed current behavior)
3. accept a callable to `fmt` so you would do `ax.bar_label(..., fmt="{:g}".format)`  (we can reliable tell apart a str and a callable)

I think option 3 is my favorite followed by 2.


----


My understand is that "fstring" refers to when you do `f"xy {code}"` which is effectivly a string literal and can not exist un-resolved where as this is about "new style format strings".

story645 commented:
I think allowing a callable would be great as that would be consistent with the tick formatters and provides the most labeling flexibility, so if that's the easiest path forward great!

stefmolin commented:
Going off the other issue, I'm going to update this to accept callables as well. I really like the idea of being able to use the tick formatters here as well.

story645 commented:
@stefmolin just to be clear, you're 💯% welcome to join in on the formatting discussion where ever it's taking place.

tacaswell commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -2745,7 +2746,22 @@ def sign(x):
+                if callable(fmt):
+                    formatted_value = fmt(value)
+                elif isinstance(fmt, str):
+                    if fmt.count('{:') == fmt.count('}') == 1:
```
This will fail to format `{{{:d}}}` which is a valid new-style format string that is not too unreasonable for a user to want.

timhoffm replied:
This is a classical case where *it's better to ask for forgiveness than for permission*.

It's probably a good idea to extract a function and put it into `matplotlib.cbook`, so that we can a) isolatedly test all kinds of input and b) reuse this later.

```
def _auto_format_str(fmt, value):
    """
    Apply *value* to the format string *fmt*.

    This works both with unnamed %-style formatting and
    unnamed {}-style formatting. %-style formatting has priority.
    If *fmt* is %-style formattable that will be used. Otherwise,
    {}-formatting is applied. Strings without formatting placeholders
    are passed through as is.

    Examples
    ---------
    >>> _auto_format_str('%.2f m', 0.2)
    '0.20 m'
    >>> _auto_format_str('{} m', 0.2)
    '0.2 m'
    >>> _auto_format_str('const', 0.2)
    'const'
    >>> _auto_format_str('%d or {}', 0.2)
    '0 or {}'
    """
    try:
        return = fmt % value
    except TypeError:
        return = fmt.format(value)
```

stefmolin replied:
I keep having an issue where the function tested on its own works, but the very same case passed into the function from `Axes.bar_label()` ends up not raising a `TypeError` when attempting %-style formatting on a {}-style format string like `{:,.0f}`. For this reason, I had to add an additional check that after attempting `fmt % value` that the result wasn't equivalent to `fmt`. I don't understand why that happens though.

story645 replied:
I'm probably wrong, but could `_preprocess_data` be messing with the calls stack in some way here 
https://github.com/matplotlib/matplotlib/blob/1d5221b2c5fe989fa6e7eb680839a1586cfa4d1e/lib/matplotlib/__init__.py#L1348

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -2622,8 +2622,9 @@ def bar_label(self, container, labels=None, *, fmt="%g", label_type="edge",
-        fmt : str, default: '%g'
-            A format string for the label.
+        fmt : str or callable, default: '%g'
+            A format string for the label or a function to call with the value
```
Please add more detailed documentation on the format variants.

timhoffm reviewed (commented):
The format function needs thorough testing.

tacaswell commented:
It is worth always trying both `%` and `.format` and warning (raising) if both work?

To @jklymak 's point about being able to force the users preference, I think "use a lambda" is sufficient for that case.

timhoffm commented:
> It is worth always trying both `%` and `.format` and warning (raising) if both work?

I originally also thought that way, but now think otherwise
- `"...".format()` returns the string unchanged if  there is no `{}` in the string. I.e. for a simple %-style string. {} works as well. - You could test that by checking whether the returned string is different, but that is extra effort.
- by prefering %-style, we have 100% backward compatibility. - That's what we have been doing so far. We only catch the exception that was earlier surfaced to the user, and now try something different. In turn, pathological cases like "%d {}" have worked so far and would break if we raise on both working.

tacaswell commented:
@timhoffm I'm convinced.

@stefmolin What strings were passing the `%` formatting unexpectedly?  That is an awkward construction of re-raising, it would be good to understand exactly why we need it....

stefmolin commented:
> @stefmolin What strings were passing the `%` formatting unexpectedly? That is an awkward construction of re-raising, it would be good to understand exactly why we need it....

@tacaswell - `{:,.0f}` was one example where the labels became `{:,.0f}`. I noticed that anything like `{:.2%}` didn't need the check &ndash; for obvious reasons. Here's a complete example that needed that check to work:

```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
bar_container = ax.barh(['c', 'b', 'a'], [1_000, 5_000, 7_000])
ax.bar_label(bar_container, label_type='center', fmt='{:,.0f}')
```

<img width="363" alt="Screen Shot 2022-08-26 at 1 25 13 PM" src="https://user-images.githubusercontent.com/24376333/186959275-51e80479-4dc4-4cc1-8fb4-5f480a625c7d.png">

tacaswell commented:
I can reproduce this 🤯 , but I think I have an explanation: 

If you put a breakpoint in the formatting function you get to:

```
fmt='{:,.0f}'
here fmt='{:,.0f}' value=1000.0 fmt % value='{:,.0f}'
> /home/tcaswell/source/p/matplotlib/matplotlib/lib/matplotlib/cbook/__init__.py(2341)_auto_format_str()
-> return fmt % value
(Pdb) fmt
'{:,.0f}'
(Pdb) value
1000.0
(Pdb) fmt % value
'{:,.0f}'
(Pdb) fmt % 5
*** TypeError: not all arguments converted during string formatting
(Pdb) fmt % value
'{:,.0f}'
(Pdb) fmt % 1000.0
*** TypeError: not all arguments converted during string formatting
(Pdb) 1000.0 == value
True
(Pdb) fmt % value
'{:,.0f}'
(Pdb) fmt % 1000.0
*** TypeError: not all arguments converted during string formatting
(Pdb) type(value)
<class 'numpy.float64'>
```

so I think the problem here is that if the value is a numpy float the string formatting fails to fail.  However it is not that numpy floats fully escape formatting:

```python
In [16]: 's' % np.float64(5)
Out[16]: 's'

In [17]: '%d' % np.float64(5)
Out[17]: '5'
```

@melissawm have you seen anything like this in numpy before?  This seems like a bug, but is rather hard to google...

tacaswell commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -2312,3 +2312,35 @@ def _unpack_to_numpy(x):
+    '0 or {}'
+    """
+    try:
+        lbl = fmt % value
+        # when used in `Axes.bar_label()`` this doesn't always raise an error
+        # when the {}-style formatting should be used instead of %-style
+        if lbl == fmt:
+            raise TypeError
+        return lbl
```
```suggestion
        return fmt % (value,)
```

I'll leave a comment explaining what is going on here.

tacaswell commented:
Ok, thanks pointers from @melissawm and Sebastian, I think I understand what is going on here:

Parsing the documnetation on this very carefully:

> If format requires a single argument, values may be a single non-tuple object. 5 Otherwise, values must be a tuple with exactly the number of items specified by the format string, or a single mapping object (for example, a dictionary).

The input can be:

 1. a single value if the format require a single argument 
 4. a tuple with exactly the right number of values
 3. a single mapping (which despite the example is really anything with `__getitem__`)

If we pass a format string with 1 `%` and our value is no a tuple, we hit case 1 and everything is copacetic.  However, if we pass a string without any `%` in it then we try the next two cases.  If it is a tuple (and the c code does a type check) then we get case 2.  If the value passed is also not a string but does have `__getitem__` then we get into case 3 where it is forgiving of some keys in the mapping not being puled out (which makes sense for a use-case where the library provides a whole lot of values and the user can pass in a constructed format string to use the (possibly empty) subset they want.  

Because `np.float64` has `__getitem__` defined we end up in case 3 unexpectedly and hence have this issue.  By putting the value (which we expect to be treated as a scalar because we told the users to provide us with a string to format a single number) we force case 2. 

Other cases that are surprising:

```python
>>> 'd' % [1, 2, 3]
'd'
>>> 'd %(0)d' % [1, 2]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: list indices must be integers or slices, not str
>>> 'd %(0)d' % {0: 1}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: '0'
>>> import numpy as np
>>> 'd %(0)d' % np.float64(5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: invalid index to scalar variable.
>>> 'd %(0)d' % np.array([1])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices
>>> 'd' % np.array([1])
'd'
>>> 'd' % {0:1}
'd'
```

The `%` formatting code in CPython https://github.com/python/cpython/blob/29f1b0bb1ff73dcc28f0ca7e11794141b6de58c9/Objects/unicodeobject.c#L14262-L14389 (grepping for the error message worked very well here) and the definition of `PyMapping_Check` https://github.com/python/cpython/blob/29f1b0bb1ff73dcc28f0ca7e11794141b6de58c9/Objects/abstract.c#L2293-L2300

stefmolin commented:
Should I add a what's new entry and/or example for this?

story645 commented:
yes what's new entry, I think update https://matplotlib.org/devdocs/gallery/lines_bars_and_markers/bar_label_demo.html#sphx-glr-gallery-lines-bars-and-markers-bar-label-demo-py with the new formatting? (use callables for one, string for another maybe?)

tacaswell commented on lib/matplotlib/tests/test_cbook.py:
```diff
@@ -895,3 +895,18 @@ def test_safe_first_element_with_none():
+])
+def test_auto_format_str(fmt, value, result):
+    """Apply *value* to the format string *fmt*."""
+    assert cbook._auto_format_str(fmt, value) == result
```
```suggestion
    assert cbook._auto_format_str(fmt, value) == result
    assert cbook._auto_format_str(fmt, np.float64(value)) == result
```

To make sure someone does not simplify the code and drop the tuple in the helper!

stefmolin commented:
> yes what's new entry, I think update either https://matplotlib.org/devdocs/gallery/lines_bars_and_markers/bar_label_demo.html#sphx-glr-gallery-lines-bars-and-markers-bar-label-demo-py or https://matplotlib.org/devdocs/gallery/lines_bars_and_markers/barchart.html#sphx-glr-gallery-lines-bars-and-markers-barchart-py with the new formatting? (use callables for one group, strings for the other group maybe?)

@story645 &ndash; I added a what's new entry, but I'm thinking it might be better to add the examples from the what's new entry to the [Bar Label Demo](https://matplotlib.org/devdocs/gallery/lines_bars_and_markers/bar_label_demo.html#sphx-glr-gallery-lines-bars-and-markers-bar-label-demo-py). The demo already has multiple examples, but it feels a bit contrived to change the current ones to use the new functionality since there is really no benefit to doing so (there's currently only one example that even uses `fmt`). What do you think?

For reference, here are the examples I created to make it more apparent why you might use the `fmt` argument in the new way:
```python
import matplotlib.pyplot as plt

animal_names = ['Lion', 'Gazelle', 'Cheetah']
mph_speed = [50, 60, 75]

fig, ax = plt.subplots()
bar_container = ax.bar(animal_names, mph_speed)
ax.set(ylabel='speed in MPH', title='Running speeds', ylim=(0, 80))
ax.bar_label(
    bar_container, fmt=lambda x: '{:.1f} km/h'.format(x * 1.61)
)
```

<img width="384" alt="Screen Shot 2022-09-03 at 9 20 10 PM" src="https://user-images.githubusercontent.com/24376333/188293062-318d1160-995c-43c3-8b32-f042cb200127.png">

```python
import matplotlib.pyplot as plt

fruit_names = ['Coffee', 'Salted Caramel', 'Pistachio']
fruit_counts = [4000, 2000, 7000]

fig, ax = plt.subplots()
bar_container = ax.bar(fruit_names, fruit_counts)
ax.set(ylabel='pints sold', title='Gelato sales by flavor', ylim=(0, 8000))
ax.bar_label(bar_container, fmt='{:,.0f}')
```

<img width="393" alt="Screen Shot 2022-09-03 at 9 23 37 PM" src="https://user-images.githubusercontent.com/24376333/188293123-4cded7b7-f5b9-4134-b9ba-2d5ca4f5c51a.png">

timhoffm commented:
> The demo already has multiple examples, but it feels a bit contrived to change the current ones to use the new functionality since there is really no benefit to doing so (there's currently only one example that even uses fmt). What do you think?

IMHO the [bar_label demo](https://matplotlib.org/devdocs/gallery/lines_bars_and_markers/bar_label_demo.html#sphx-glr-gallery-lines-bars-and-markers-bar-label-demo-py) needs an overhaul anyway. The semanitc content in the existing examples is quite far fetched and doesn't really make sense. I'd like to see more realistic content, like your plots.

For now, it's fine to add your plots to the bar label demo as is.

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -2622,8 +2622,10 @@ def bar_label(self, container, labels=None, *, fmt="%g", label_type="edge",
+        fmt : str or callable, default: '%g'
+            An unnamed %-style or {}-style format string for the label or a
+            function to call with the value as the first argument.
+            When *fmt* is a string, %-style takes precedence over {}-style.
```
Let's make it very explicit when this is relevant.
```suggestion
            When *fmt* is a string and can be interpreted in both formats,
            %-style takes precedence over {}-style.
```

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -2745,7 +2747,12 @@ def sign(x):
 
-            annotation = self.annotate(fmt % value if lbl is None else lbl,
+            if lbl is None:
+                try:
+                    lbl = fmt(value)
+                except TypeError:
+                    lbl = cbook._auto_format_str(fmt, value)
```
```suggestion
                if isinstance(fmt, str):
                    lbl = cbook._auto_format_str(fmt, value)
                elif callable(fmt):
                    lbl = fmt(value)
                else:
                    raise TypeError("fmt must be a str or callable")
```
In this case we should do explicit checks:

- If `fmt()` is callable but produces a TypeError internally, the try-except version would fall back to the str code, which would not work and create a strange error message.
- With only two additional lines, we can add a proper error message, if the input type is unsupported.

Side-note: I have put the str-test first because that's the most-common (and default) usage. The code is more readable when you go in the order common-less_used-error in if chains.

timhoffm reviewed (approved):
I still have two minor suggestions, but approve already what is there.

timhoffm commented:
Off-topic: @stefmolin Thanks for working on Matplotlib. I'd like to make you aware of the [new contributor meeting](https://hackmd.io/TYTHxxvSRWugPCPEOox2gA). This is a good place to get to know some of the matplotlib developers, to learn more about the project and of course to ask questions in general. If you are interested, you are very welcome to join. The next date is Tuesday, Sept 6 at 5pm UTC.

stefmolin commented:
@timhoffm - I've incorporated the suggestions and added the two new examples to the demo. Thanks for letting me know about the new contributor meeting &ndash; it's during work for me, but I'll see if I can attend 😄

jklymak commented:
It is milestoned 3.7, so shouldn't be a problem, right?

timhoffm commented:
> Yeah was just unsure if feature freeze comes in once the RCs are cut or if they can come in - under the wire.

3.6.0 feature freeze comes in as soon as we have the 3.6.x branch. Then, the PRs are going into main and need to be backported to the 3.6.x brach explicitly.
